Src/library/net/Clib/network.c

```diff
@@ -1189,32 +1189,44 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 	int arg;
 	asize = sizeof(arg);
 	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_RCVTIMEO, (char *) &arg, &asize));
-	return (EIF_INTEGER) (arg / 1000); /* Convert milliseconds to seconds */
+	return (EIF_INTEGER) arg; 
 #else
 	struct timeval tv;
 	asize = sizeof(struct timeval);
 	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_RCVTIMEO, &tv, &asize));
-	return (EIF_INTEGER) tv.tv_sec;
+	return (EIF_INTEGER) (1000 * tv.tv_sec + tv.tv_usec / 1000);
 #endif
 }
 
-void c_set_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER recv_timeout_seconds)
-	/* set socket SO_RCVTIMEO option to recv_timeout_seconds */
+EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+	/* get socket fd SO_RCVTIMEO options in seconds */
+{
+	return (EIF_INTEGER) (c_get_sock_recv_timeout_ms(fd, level) / 1000); /* Convert milliseconds to seconds */
+}
+
+void c_set_sock_recv_timeout_ms(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER recv_timeout_milliseconds)
+	/* set socket SO_RCVTIMEO option to recv_timeout_milliseconds */
```

Comment:
On Windows ... the C API accepts milliseconds, not usecs.

I made this patch a while ago .. but after thinking again at it
we should mimic routine for timeout (select_poll timeout) ..
to use 2 int values, one for the seconds, and another for usecs.
This way, if in the future, Windows accepts usecs, the Eiffel interface won't change, only the EiffelNet C code.
so it would be

void c_set_sock_recv_timeout_ms(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER recv_timeout_seconds, EIF_INTEGER timeout_usec) {..}

Reply:
Some other socket API support nanoseconds - check recvmmsg, pselect, ppoll, etc. - they use nanoseconds for timeouts. Therefore I would go for Eiffel API that uses nanoseconds everywhere. Otherwise we'll have seconds, microseconds, milliseconds, nanoseconds - too many different time units. Seconds plus nanoseconds should be enough for everything.

Moreover, we can have a special Eiffel type that encapsulates those values and allows working on a higher level of abstraction.

Reply:
I agree after all.